underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -421,30 +422,39 @@ protected ObjectListingChunk getObjectListingChunk(String key, boolean recursive
   // Get next chunk of listing result.
   private ListObjectsV2Result getObjectListingChunk(ListObjectsV2Request request)
       throws IOException {
-    ListObjectsV2Result result;
-    try {
-      // Query S3 for the next batch of objects.
-      result = mClient.listObjectsV2(request);
-      // Advance the request continuation token to the next set of objects.
-      request.setContinuationToken(result.getNextContinuationToken());
-    } catch (AmazonClientException e) {
-      throw new IOException(e);
+    // S3A write-then-list workload may has problems due to its eventual consistency issue
+    RetryPolicy retryPolicy = getRetryPolicy();
+    IOException thrownException = null;
+    while (retryPolicy.attempt()) {
+      try {
+        // Query S3 for the next batch of objects.
+        ListObjectsV2Result result = mClient.listObjectsV2(request);
+        // Advance the request continuation token to the next set of objects.
+        request.setContinuationToken(result.getNextContinuationToken());
+        return result;
+      } catch (AmazonClientException e) {
+        thrownException = new IOException(e);
+      }
     }
-    return result;
+    throw thrownException;
   }
 
   // Get next chunk of listing result.
   private ObjectListing getObjectListingChunkV1(ListObjectsRequest request) throws IOException {
-    ObjectListing result;
-    try {
-      // Query S3 for the next batch of objects.
-      result = mClient.listObjects(request);
-      // Advance the request continuation token to the next set of objects.
-      request.setMarker(result.getNextMarker());
-    } catch (AmazonClientException e) {
-      throw new IOException(e);
+    RetryPolicy retryPolicy = getRetryPolicy();
+    IOException thrownException = null;
+    while (retryPolicy.attempt()) {
```

Comment:
can we refactor this common retry pattern? Something like

```java
private interface AmazonClientOperation<T> {
  T apply() throws AmazonClientException;
}

private <T> T retry(AmazonClientOperation<T> op) throws AmazonClientException {
  
}
```

Reply:
Refactored, PTAL, thanks